scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala

```diff
@@ -365,207 +422,93 @@ class FormatWriter(formatOps: FormatOps) {
-            !left.is[T.LeftParen] && // skip empty parentheses
-            !formatToken.right.is[T.Comment] &&
-            rightIsCloseDelim && isNewline =>
-        sb.append(",")
-        sb.append(whitespace)
-
-      // foo(
-      //   a,
-      //   b // comment
-      // )
-      //
-      // Insert a comma after b (before comment)
-      case TrailingCommas.always
-          if left.is[T.Comment] && !prevFormatToken.left.is[T.Comma] &&
-            !prevFormatToken.left.is[T.Comment] &&
-            !prevNonComment(formatToken).left
-              .is[T.LeftParen] && // skip empty parentheses
-            rightIsCloseDelim && isNewline =>
-        val indexOfComment = sb.lastIndexOf(left.syntax)
-        val index = sb.lastIndexOf(prevFormatToken.left.syntax, indexOfComment)
-
-        // If the leading comment is vertically aligned, preserve the location of
-        // the comment to avoid breaking the alignment.
-        if (tokenAligns.get(prevFormatToken).isDefined) {
-          sb.setCharAt(index + prevFormatToken.left.syntax.length, ',')
-        } else {
-          sb.insert(index + prevFormatToken.left.syntax.length, ",")
-        }
-        sb.append(whitespace)
-
-      // foo(
-      //   a,
-      //   b,
-      // )
-      //
-      // Remove the comma after b
-      case TrailingCommas.never
-          if left.is[T.Comma] && rightIsCloseDelim &&
-            !formatToken.right.is[T.Comment] && isNewline =>
-        sb.deleteCharAt(sb.length - 1)
-        sb.append(whitespace)
-
-      // foo(
-      //   a,
-      //   b, // comment
-      // )
-      //
-      // Remove the comma after b (before comment)
-      case TrailingCommas.never
-          if left.is[T.Comment] && prevFormatToken.left.is[T.Comma] &&
-            rightIsCloseDelim && isNewline =>
-        val indexOfComment = sb.lastIndexOf(left.syntax)
-        val indexOfComma =
-          sb.lastIndexOf(prevFormatToken.left.syntax, indexOfComment)
-
-        // If the leading comment is vertically aligned, preserve the location of
-        // the comment to avoid breaking the alignment.
-        if (tokenAligns.get(prevFormatToken).isDefined) {
-          sb.setCharAt(indexOfComma, ' ')
-        } else {
-          sb.deleteCharAt(indexOfComma)
-        }
-        sb.append(whitespace)
-
-      // foo(a, b,)
-      //
-      // Remove the comma after b
-      case _
-          if left.is[T.Comma] && rightIsCloseDelim &&
-            !formatToken.right.is[T.Comment] && !isNewline =>
-        sb.deleteCharAt(sb.length - 1)
-
-      case _ => sb.append(whitespace)
-    }
-
-  }
 }
 
 object FormatWriter {
 
   case class FormatLocation(
       formatToken: FormatToken,
-      split: Split,
       state: State
   )
+
+  private val indentations: Array[String] = {
+    val arr = new Array[String](64)
+    arr(0) = ""
+    (1 until arr.length).foreach(idx => arr(idx) = " " + arr(idx - 1))
+    arr
+  }
+
+  private def getIndentation(len: Int): String =
```

Comment:
Can you please write a comment with explanation of this trick? I didn't immediately realize that it is string concatenation cache

Reply:
will do

Reply:
will do